src/pmt.c

```diff
@@ -659,14 +659,28 @@ int fill_packet_start_indicator(SPMT_batch *all, int max_all, SPMT_batch *start,
     return s;
 }
 
+void dump_cws() {
+    int i;
+    char buf[200];
+    uint64_t ctime = getTick();
+    LOG("List of CWs:");
+    for (i = 0; i < ncws; i++)
+        if (cws[i] && cws[i]->enabled && cws[i]->expiry > ctime) {
+            LOG("* CW %d: %s", cws[i]->id, cw_to_string(cws[i], buf));
+        }
+}
+
 void update_cw(SPMT *pmt) {
     SCW *cw = NULL, *old_cw = pmt->cw;
     char buf[300];
     int64_t ctime = getTick();
     int i = 0;
     SPMT_batch start[10];
     int len = fill_packet_start_indicator(pmt->batch, pmt->blen, start, 9);
-    int validated = (len > 1);
+    int validated = 1;
+
+    if (len <= 0)
+        validated = 0;
```

Comment:
Couldn't all this be simplified into `int validated = (len > 0);`?

Reply:
> Couldn't all this be simplified into `int validated = (len > 0);`?


No. That's the reason for the patch... As I say in the description **_several compilers fail to initialize correctly_**. You can check it in your system activating the log with `-v pmt`. If you see multiple times that the same key is checked then your compiler has troubles. So please, belive me: it's preferable writing the code in this way. The result never fails, and if the compiler wants to optimize it, it can do it. Therefore this patch doesn't introduce any penalty in the performance.

I've dedicated a lot of time to found this!
My recomendation is to merge this PR now.

Reply:
Have you tried `volatile int validated = (len > 0);`?

Reply:
The thing is you're both changing the logic _and_ adding some form of optimization hack. Are you sure it's not enough to just change the logic?

Reply:
Hi @Jalle19 (and @catalinii ),

Sorry, sorry! I revised the code, and the problem is another. The correct solution is the one you proposed:
```
int validated = (len > 0);
```
So, yes. In the original code some bug exist because with only 1 packet with the PUSI flag it's sufficient to do the validation. I don't know why the original author used `validated = (len > 1)` as it doesn't have sense.

Now, it's fixed, so please, merge the PR.
Regards.

Reply:
Very good, thanks for checking! Now the fix makes perfect sense